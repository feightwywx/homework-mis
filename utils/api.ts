import { NextApiRequest } from "next";
import { JsonResponse, ResponseResult, UserType } from "./types";
import { getId } from "./user";

export function successResponse<T = ResponseResult>(result?: T) {
  return {
    code: 0, 
    result: result ?? ''
  } as JsonResponse<T>
}

export function failResponse(code: number, result?: ResponseResult) {
  return {
    code, 
    result: result ?? ''
  } as JsonResponse
}

export async function parseIdFromReqest(req: NextApiRequest, userType: UserType) {
  let token = undefined;
  if (req.session.user?.token) {
    token = req.session.user?.token
  } else {
    token = await req.body.token;
  }

  return await getId(userType, token);
}

export const statusCode = {
  ROUTER_PARAM_REQUIRED: 100,
  ROUTER_PARAM_INVALID: 1001,
  BODY_PARAM_REQUIRED: 101,
  URL_PARAM_REQUIRED: 102,
  TOKEN_INVALID: 200,
  NUL_QUERY_DATA: 300,
  NUL_QUERY_PARAM: 301,
  BAD_UPDATE: 400
}
