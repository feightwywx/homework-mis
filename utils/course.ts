import { createMisConnection } from "./mysql";
import { Course } from "./types";

type CourseRow = {
  id: number;
  name: string;
  time: string;
  ended: number;
  teacherID: number;
  teacherName: string;
}

export async function getStudentCourses(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.* , teacher.name AS teacherName
    FROM course
    JOIN course_student ON course.id = course_student.studentID 
    JOIN teacher ON course.teacherID = teacher.id
    WHERE course_student.studentID = ?;`,
    [id]
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses;
}

export async function getTeacherCourses(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.* , teacher.name AS teacherName
    FROM course
    JOIN teacher ON course.teacherID = teacher.id
    WHERE teacher.id = ?;`,
    [id]
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses;
}

export async function getCourseByID(id: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.* , teacher.name AS teacherName
    FROM course
    JOIN teacher ON course.teacherID = teacher.id
    WHERE course.id = ?;`,
    [id]
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses[0];
}

export async function getCourses() {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT course.*
    FROM course`
  );
  await conn.end();

  const courses = (rows as Array<CourseRow>).map((row) => {
    return {
      ...row,
      ended: row.ended === 0 ? false : true,
    } as Course;
  });

  return courses;
}
