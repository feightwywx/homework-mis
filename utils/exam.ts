import { ResultSetHeader } from "mysql2";
import { createMisConnection } from "./mysql";
import { Exam, ExamResult } from "./types";

type ExamRow = {
  id: number;
  name: string;
  time: string;
  endtime: string;
  location: string;
  courseID: number;
};

type ExamResultRow = {
  score: number | null;
};
type DetailedExamResultRow = ExamResultRow & {
  id: number;
  studentID: number;
  studentName: string;
};

export async function getExamsByCourseID(coid: number): Promise<Exam[]> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT *
    FROM exam
    WHERE courseID=?`,
    [coid]
  );
  await conn.end();

  const exams = rows as Array<ExamRow>;

  return exams;
}

export async function getExamByID(eid: number): Promise<Exam> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT *
    FROM exam
    WHERE id=?`,
    [eid]
  );
  await conn.end();

  const exams = (rows as Array<ExamRow>)[0];

  return exams;
}

export async function getStudentExamScore(
  uid: number,
  eid: number
): Promise<number | null> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT score
    FROM exam_result
    WHERE studentID=? AND examID=?;`,
    [uid, eid]
  );
  await conn.end();

  const score = (rows as Array<ExamResultRow>)[0].score;
  console.log("score", score);

  return score;
}

export async function getTeacherExamScore(eid: number): Promise<ExamResult[]> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `
    SELECT exam_result.id, studentID, student.name AS studentName, score
    FROM exam_result
    JOIN student ON student.id = exam_result.studentID
    WHERE exam_result.examID = ?;`,
    [eid]
  );
  await conn.end();

  const score = rows as Array<DetailedExamResultRow>;

  return score;
}

export async function updateExamScore(
  erid: number,
  score: number
): Promise<number> {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    `UPDATE exam_result SET score=? WHERE id=?;`,
    [score, erid]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows;
}

export async function insertExamAssign(courseID: number, title: string, location: string, time: string, endtime: string) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'INSERT INTO exam (name, location, time, endtime, courseID) ' +
    'VALUES (?, ?, ?, ?, ?) ',
    [title, location, time, endtime, courseID]
  );
  await conn.end();

  return (rows as ResultSetHeader).insertId
}

export async function insertExamTarget(studentID: number, examID: number) {
  const conn = await createMisConnection();
  const [rows] = await conn.query(
    'INSERT INTO exam_result (studentID, examID) ' +
    'VALUES (?, ?) ',
    [studentID, examID]
  );
  await conn.end();

  return (rows as ResultSetHeader).affectedRows
}
