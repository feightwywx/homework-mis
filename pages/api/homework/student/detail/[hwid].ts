import { withIronSessionApiRoute } from "iron-session/next";
import { NextApiRequest, NextApiResponse } from "next";
import { getStudentHomeworkDetail } from "../../../../../utils/homework";
import { failResponse, parseIdFromReqest, statusCode, successResponse } from "../../../../../utils/api";
import { sessionOptions } from "../../../../../utils/session";

export async function studentDetailRoute(req: NextApiRequest, res: NextApiResponse) {
  const { hwid } = req.query;

  if (!hwid) {
    res.json(failResponse(statusCode.ROUTER_PARAM_REQUIRED, 'hwid required'));
    return;
  }

  const id = await parseIdFromReqest(req, 'student');
  if (!id) {
    res.json(failResponse(statusCode.TOKEN_INVALID));
    return;
  }

  const result = await getStudentHomeworkDetail(+hwid, id);
  if (result) {
    res.json(successResponse(result));
  } else {
    res.json(failResponse(statusCode.NUL_QUERY_DATA));
  }
  return;
}

export default withIronSessionApiRoute(studentDetailRoute, sessionOptions)
