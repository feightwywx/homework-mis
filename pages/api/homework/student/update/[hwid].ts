import { withIronSessionApiRoute } from "iron-session/next";
import { NextApiRequest, NextApiResponse } from "next";
import { updateHomework } from "../../../../../utils/homework";
import { sessionOptions } from "../../../../../utils/session";
import { getId } from "../../../../../utils/user";

async function studentUpdateRoute(req: NextApiRequest, res: NextApiResponse) {
  const { hwid } = req.query;
  const { content } = await req.body;

  if (!hwid) {
    res.status(500).end();
    return;
  }

  let token = undefined;
  if (req.session.user?.token) {
    token = req.session.user?.token
  } else {
    token = await req.body.token;
  }

  const id = await getId('student', token);

  if (!id) {
    res.status(401).end();
    return;
  }

  const result = await updateHomework(+hwid, id, content);
  res.json({success: result});
}

export default withIronSessionApiRoute(studentUpdateRoute, sessionOptions)
