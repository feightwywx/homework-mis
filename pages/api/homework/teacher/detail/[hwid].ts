import { withIronSessionApiRoute } from "iron-session/next";
import { NextApiRequest, NextApiResponse } from "next";
import { getTeacherHomeworkDetail } from "../../../../../utils/homework";
import { sessionOptions } from "../../../../../utils/session";
import { getId } from "../../../../../utils/user";

async function teacherDetailRoute(req: NextApiRequest, res: NextApiResponse) {
  const { hwid } = req.query;

  if (!hwid) {
    res.status(500).end();
    return;
  }

  let token = undefined;
  if (req.session.user?.token) {
    token = req.session.user?.token
  } else {
    token = await req.body.token;
  }

  const id = await getId('teacher', token);

  if (!id) {
    res.status(401).end();
    return;
  }

  const result = await getTeacherHomeworkDetail(+hwid, id);
  res.json(result);
}

export default withIronSessionApiRoute(teacherDetailRoute, sessionOptions)
