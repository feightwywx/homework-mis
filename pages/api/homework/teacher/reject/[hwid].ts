import { withIronSessionApiRoute } from "iron-session/next";
import { NextApiRequest, NextApiResponse } from "next";
import { failResponse, parseIdFromReqest, statusCode, successResponse } from "../../../../../utils/api";
import { setReject } from "../../../../../utils/homework";
import { sessionOptions } from "../../../../../utils/session";
import { getStudentId } from "../../../../../utils/user";

export async function teacherRejectRoute(req: NextApiRequest, res: NextApiResponse) {
  const { hwid } = req.query;
  const { studentName } = await req.body;
  const stuid = await getStudentId(studentName);

  if (!hwid) {
    res.json(failResponse(statusCode.ROUTER_PARAM_REQUIRED, 'hwid required'));
    return;
  }
  if (!studentName) {
    res.json(failResponse(statusCode.BODY_PARAM_REQUIRED, 'studentName required'));
    return;
  }
  if (!stuid) {
    res.json(failResponse(statusCode.NUL_QUERY_PARAM, 'stuid query error'));
    return;
  }

  const id = await parseIdFromReqest(req, 'teacher');
  if (!id) {
    res.json(failResponse(statusCode.TOKEN_INVALID));
    return;
  }

  const result = await setReject(+hwid, stuid);
  if (result) {
    res.json(successResponse({ affected: result }))
  } else {
    res.json(failResponse(statusCode.NUL_QUERY_DATA))
  }
  return;
}

export default withIronSessionApiRoute(teacherRejectRoute, sessionOptions)
