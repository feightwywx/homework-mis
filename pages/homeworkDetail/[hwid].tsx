import { NextPage } from 'next';
import { useRouter } from 'next/router';
import HwLayout from '../../components/layout';
import { Typography, Space, Button, Divider, Tag, Modal, Input, Table, Form, InputNumber } from 'antd';
import { ArrowLeftOutlined, FormOutlined } from '@ant-design/icons';
import { useMediaPredicate } from "react-media-hook";
import useUser from '../../utils/hooks/useUser';
import { HomeworkDetail, HomeworkDetailContent, HomeworkStudentDetail, HomeworkTeacherDetail, HomeworkTeacherDetailContent } from '../../utils/types';
import React, { useState } from 'react';
import useSWR from 'swr';
import parseMysqlDateTime from '../../utils/parseTime';
import { ColumnsType } from 'antd/lib/table';

const { Title, Text, Paragraph } = Typography;
const { TextArea } = Input;

function HomeworkDetailPage() {
  const router = useRouter();
  const { hwid } = router.query;

  const { user } = useUser();

  const { data } = useSWR(`/api/homework/${user?.userType}/detail/${hwid}`);

  const detail = (data as HomeworkStudentDetail)?.detail;
  const content = (data as HomeworkStudentDetail)?.content;
  const dead = detail ? new Date(Date.now()) > parseMysqlDateTime(detail.deadline) : false;

  return (
    <HwLayout>
      {data && hwid &&
        <>
          <CommonDetail content={detail} />
          {user?.userType === 'student'
            ? <StudentDetail content={content} dead={dead} hwid={+hwid} />
            : <TeacherDetail content={(data as HomeworkTeacherDetail).content} hwid={+hwid} />
          }

        </>
      }
    </HwLayout>
  )
}

export function CommonDetail({ content: detail }: { content: HomeworkDetail }) {
  const router = useRouter();

  return (<>
    <Space direction="vertical" size="middle" style={{ display: 'flex' }}>
      <div>
        <Button
          icon={<ArrowLeftOutlined />}
          style={{
            float: 'left', verticalAlign: 'middle', margin: '8px 24px 0 0'
          }}
          onClick={() => { router.push('/') }}
        />
        <Title style={{ float: 'left' }}>{detail.title}</Title>
      </div>
      <Space direction='vertical' size='small'>
        <Text>发布时间：{parseMysqlDateTime(detail.time).toLocaleString()}</Text>
        <Text>截止时间：{parseMysqlDateTime(detail.deadline).toLocaleString()}</Text>
      </Space>
      <Divider />
      <Title level={2}>作业内容</Title>
      <Paragraph>
        {detail.assignment}
      </Paragraph>
      <Divider />
    </Space>
  </>)
}

export function StudentDetail({
  content, dead = false, hwid, teacher = false
}: {
  content: HomeworkDetailContent | HomeworkTeacherDetailContent,
  dead?: boolean,
  hwid: number,
  teacher?: boolean
}) {
  const [modalOpen, setModalOpen] = useState(false);
  const [accContent, setAccContent] = useState('');
  const [confirmLoading, setConfirmLoading] = useState(false);
  
  const router = useRouter()

  function accFieldHandler(e: React.ChangeEvent<HTMLTextAreaElement>) {
    setAccContent(e.target.value);
  }

  function submitHandler() {
    setConfirmLoading(true);
    fetch(`/api/homework/student/update/${hwid}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ content: accContent })
    }).then(() => {
      setConfirmLoading(false);
      setModalOpen(false);
      router.reload();
    }).catch(err => {
      console.error(err)
    })
  }

  const matches = useMediaPredicate("(min-width: 768px)")

  return (
    <>
      <Space direction="vertical" size="middle" style={{ display: 'flex' }}>
        <div>
          <Title level={2} style={{ float: 'left' }}>
            {!teacher ? '我的提交' : `${(content as HomeworkTeacherDetailContent).studentName}的提交内容`}
          </Title>
          {
            !teacher &&
            <Button
              type='primary'
              shape='round'
              icon={<FormOutlined />}
              style={{ float: 'right', marginTop: 8 }}
              disabled={content.completed || dead}
              onClick={() => { setModalOpen(true); }}
            >
              提交作业
            </Button>
          }

        </div>
        {(!content.completed && !content.accomplishment?.content && !dead)
          && <Paragraph>{!teacher ? '目前还没有提交过这项作业。点击右上角的按钮来提交吧。' : '该学生没有提交这项作业。'}</Paragraph>}
        {(!content.completed && !content.accomplishment?.content && dead)
          && <Paragraph>{!teacher ? '这项作业已经过期，无法提交！': '该学生没有提交这项作业。'}</Paragraph>}
        {content.accomplishment?.content
          && <div>
            <Title level={3} style={{ float: 'left', marginBottom: 0 }}>提交内容</Title>
            <Paragraph style={{ float: 'right', marginTop: 6 }}>
              {matches && '提交时间：'}{parseMysqlDateTime(content.accomplishment.time).toLocaleString()}
            </Paragraph>
          </div>}
        <Paragraph>
          {/* FIXME: 换行渲染不出来 */}
          {content.accomplishment?.content}
        </Paragraph>
        {content.accomplishment?.content &&
          <>
            <Divider />
            <Title level={2}>作业评价</Title>
            {content.judge ?
              <>
                <div>
                  <Title level={3} style={{ float: 'left' }}>得分</Title>
                  <Text style={{ float: 'right', fontSize: 18 }}>/100</Text>
                  <Text
                  style={{ float: 'right', fontSize: 18 }}
                  type={content.judge.score
                    ? 'success'
                  : (content.judge.comment
                  ? 'warning'
                  : 'danger')}>
                    {content.judge.score ?? (content.judge.comment && '已打回')}
                    </Text>
                </div>
                <Title level={3} style={{ float: 'left', marginBottom: 0 }}>评语</Title>
                <Paragraph>
                  {content.judge?.comment ?
                    content.judge.comment :
                    (!teacher ? '老师没有留下评语哦～' : '没有留下评语。')}
                </Paragraph>
              </> :
              <Paragraph>{!teacher ? '老师还没有批阅你的作业，请耐心等待。' : '还没有批阅这项作业。'}</Paragraph>}
          </>}
      </Space>
      <Modal
        open={modalOpen}
        title='作业提交'
        onCancel={() => { setModalOpen(false); }}
        width={'80vw'}
        cancelText='取消'
        okText='提交'
        confirmLoading={confirmLoading}
        onOk={submitHandler}
      >
        <Paragraph>请输入需要提交的内容。</Paragraph>
        <TextArea
          rows={20}
          onChange={accFieldHandler}
          onBlur={accFieldHandler}
          showCount
        />
      </Modal>
    </>)
}

export function TeacherDetail({ content, hwid }: { content: Array<HomeworkTeacherDetailContent>, hwid: number }) {
  interface DataType {
    key: number;
    name: string;
    status: string[];
    time: string;
  }

  const TagColor = {
    '已完成': 'green',
    '待评分': 'magenta',
    '已评分': 'lime'
  } as { [x: string]: string }

  const router = useRouter();

  const [judgeModalOpen, setJudgeModalOpen] = useState(false);
  const [detailModalOpen, setDetailModalOpen] = useState(false);
  const [currentDetail, setCurrentDetail] = useState({} as HomeworkTeacherDetailContent)
  const [confirmJudgeLoading, setConfirmJudgeLoading] = useState(false);
  const [currentContentId, setCurrentContentId] = useState(0);

  const [form] = Form.useForm();

  const judgeClickHandler = (values: {
    score: number, comment?: string
  }) => {
    setConfirmJudgeLoading(true);

    fetch(`/api/homework/teacher/judge/${currentContentId}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ score: values.score, comment: values.comment })
    }).then(() => {
      router.reload();
      setConfirmJudgeLoading(false);
      setJudgeModalOpen(false);
      setDetailModalOpen(false);
    }).catch(err => {
      console.error(err)
    })
  };

  function rejectClickHandler() {
    fetch(`/api/homework/teacher/reject/${hwid}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
    }).then(() => {
      router.reload();
    }).catch(err => {
      console.error(err);
    })
  }

  return (<>
    <Space direction="vertical" size="middle" style={{ display: 'flex' }}>
      <Title level={2}>已提交</Title>
      <Table
        dataSource={content.map((item) => {
          let status = [];
          if (item.completed) {
            status.push('已完成')

            if (!item.judge || (item.judge.score === null)) {
              status.push('待评分')
            } else {
              status.push('已评分')
            }
          }

          if (!item.completed) {
            if (item.accomplishment?.content) {
              status.push('已打回')
            } else {
              status.push('未完成')
            }
          }

          return {
            key: item.contentId,
            name: item.studentName,
            status: status,
            time: item.accomplishment?.time ? parseMysqlDateTime(item.accomplishment?.time).toLocaleString() : ''
          }
        })}
        columns={[
          {
            title: '姓名',
            dataIndex: 'name',
            key: 'name'
          },
          {
            title: '状态',
            dataIndex: 'status',
            key: 'status',
            render: (_, { status }) => (
              <>
                {status.map(tag => {
                  return (
                    <Tag color={TagColor[tag] ?? 'grey'} key={tag}>
                      {tag}
                    </Tag>
                  );
                })}
              </>
            )
          },
          {
            title: '提交时间',
            dataIndex: 'time',
            key: 'time'
          },
          {
            title: '操作',
            key: 'action',
            render: (_, record) => (
              <Space size='small'>
                <Button
                  type='link'
                  onClick={() => {
                    const promise = new Promise<HomeworkTeacherDetailContent>((resolve, reject) => {
                      const curr = content.filter((item) => (item.contentId === record.key))[0]
                      if (curr) {
                        resolve(curr);
                      } else {
                        reject();
                      }

                    })
                    promise.then((data: HomeworkTeacherDetailContent) => {
                      setCurrentContentId(data.contentId)
                      setCurrentDetail(data as HomeworkTeacherDetailContent);
                      setDetailModalOpen(true);
                    })
                  }}>
                  查看
                </Button>
                <Button
                  type='link'
                  disabled={record.status.includes('未完成') || record.status.includes('已打回')}
                  onClick={() => { setJudgeModalOpen(true); setCurrentContentId(record.key); }}>
                  批阅
                </Button>
                <Button
                  type='link'
                  disabled={record.status.includes('未完成') || record.status.includes('已打回')}
                  onClick={rejectClickHandler}>
                  打回
                </Button>
              </Space>
            )
          }] as ColumnsType<DataType>} />
    </Space>
    <Modal
      open={judgeModalOpen}
      title='作业批阅'
      onCancel={() => { setJudgeModalOpen(false); }}
      width={'80vw'}
      confirmLoading={confirmJudgeLoading}
      onOk={() => {
        form
          .validateFields()
          .then(values => {
            form.resetFields();
            judgeClickHandler(values);
          })
          .catch(() => {
          });
      }}
    >
      <Form
        form={form}
        autoComplete='off'
        requiredMark='optional'>
        <Form.Item
          label='得分 　　　　'
          name='score'
          rules={[{
            required: true,
            message: '请输入得分！'
          },
          {
            type: 'number',
            min: 0,
            max: 100,
            message: '得分需要在0～100之间！'
          }]}>
          <InputNumber />
        </Form.Item>
        <Form.Item
          label='评语'
          name='comment'>
          <TextArea
            rows={10}
            showCount
          />
        </Form.Item>

      </Form>
    </Modal>
    <Modal
      open={detailModalOpen}
      title='作业详情'
      onCancel={() => { setDetailModalOpen(false); setJudgeModalOpen(false); }}
      width={'80vw'}
      onOk={() => {
        setJudgeModalOpen(true);
      }}
      okText='批阅'
      okButtonProps={{
        disabled: !currentDetail.completed
      }}
    >
      <StudentDetail content={currentDetail} hwid={1} teacher={true} />
    </Modal>
  </>)
}

export default HomeworkDetailPage as NextPage;
