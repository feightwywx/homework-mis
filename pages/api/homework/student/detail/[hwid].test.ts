import { createMocks } from 'node-mocks-http';
import type { NextApiRequest, NextApiResponse as OrigNextApiResp } from 'next';
import { studentDetailRoute } from './[hwid]';
import type { JsonResponse } from '../../../../../utils/types';
import { initTestDatabase } from '../../../../../test/utils/initTestDatabase';

interface NextApiResponse<T = any> extends OrigNextApiResp<T> {
  _getJSONData: () => T;
}

describe('/api/homework/student/detail endpoint', () => {
  beforeAll(async () => {
    return initTestDatabase()
  })

  const token = '74985a778511a53ba671334bbde2125f0033a8f16907c2d8965cac571645a96ad60333d179021e65fc65ac20cf42e009';

  function mockRequestResponse(loggedIn = false, hwid?: number) {
    // @ts-expect-error
    const { req, res }: { req: NextApiRequest, res: NextApiResponse<JsonResponse> } = createMocks({
      method: 'GET',
      body: loggedIn ? { token } : {},
    })

    if (typeof hwid === 'number') {
      req.query = { hwid: hwid.toString() };
    }
    return { req, res }
  }

  it('returns 100 if !hwid', async () => {
    const { req, res } = mockRequestResponse();
    await studentDetailRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData().code).toEqual(100);
  })

  it('returns 200 if !hwid', async () => {
    const { req, res } = mockRequestResponse(false, 1);
    await studentDetailRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData().code).toEqual(200);
  })

  it('returns homework detail', async () => {
    const { req, res } = mockRequestResponse(true, 1);
    await studentDetailRoute(req, res);

    expect(res.statusCode).toBe(200);
    expect(res._getJSONData()).toMatchSnapshot();
  })

  it('returns 300 if homework does not exist', async () => {
    const { req, res } = mockRequestResponse(true, -1);
    await studentDetailRoute(req, res);

    expect(res.statusCode).toBe(200);
    // @ts-ignore
    expect(res._getJSONData().code).toEqual(300);
  })
})