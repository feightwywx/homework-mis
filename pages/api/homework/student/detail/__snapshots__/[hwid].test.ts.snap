// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/homework/student/detail endpoint returns homework detail 1`] = `
{
  "code": 0,
  "result": {
    "content": {
      "completed": false,
    },
    "detail": {
      "assignment": "这是一个测试用的作业。用于测试未完成作业。",
      "deadline": "2023-10-30T03:45:14.000Z",
      "time": "2022-09-27T02:50:02.000Z",
      "title": "作业1",
    },
  },
}
`;
